### ISSUE TITLE
Configuration resolution is not safe and is scheduled to be removed in Gradle 7.0

### ISSUE THREAD

**Issue by vlsi** [mnlipp/jdrupes-mdoclet#4]
Hi,

It looks like the current `jdrupes-mdoclet` will be incompatible with Gradle 7.
The related issue is https://github.com/gradle/gradle/issues/11748

The workaround (== workaround "a") seems to be as follows (plugin code, in Kotlin, Apache 2.0): https://github.com/autostyle/autostyle/blob/f201199bc327887225db38ae34261bfb4a861527/buildSrc/src/main/kotlin/mdoclet.gradle.kts

What do you think if https://github.com/mnlipp/jdrupes-mdoclet/blob/50f4adc78570193f134e82c0c1dc6be09f39541b/integrations/gradle-plugin/src/org/jdrupes/mdoclet/integrations/gradle/MDocletPlugin.groovy is rewritten in Kotlin like `mdoclet.gradle.kts` above?

PS. I can provide a PR provided it is fine to rewrite `MDocletPlugin.groovy` in Kotlin.

The warnings are as follows:

> Configure project :autostyle-lib
> The configuration :autostyle-lib-extra:org.jdrupes.mdoclet was resolved without accessing the project in a safe manner.  This may happen when a configuration is resolved from a different project.  See https://docs.gradle.org/6.0.1/userguide/viewing_debugging_dependencies.html#sub:resolving-unsafe-configuration-resolution-errors for more details. This behaviour has been deprecated and is scheduled to be removed in Gradle 7.0.
> The configuration :autostyle-plugin-gradle:org.jdrupes.mdoclet was resolved without accessing the project in a safe manner.  This may happen when a configuration is resolved from a different project.  See https://docs.gradle.org/6.0.1/userguide/viewing_debugging_dependencies.html#sub:resolving-unsafe-configuration-resolution-errors for more details. This behaviour has been deprecated and is scheduled to be removed in Gradle 7.0.
> The configuration :autostyle-testlib:org.jdrupes.mdoclet was resolved without accessing the project in a safe manner.  This may happen when a configuration is resolved from a different project.  See https://docs.gradle.org/6.0.1/userguide/viewing_debugging_dependencies.html#sub:resolving-unsafe-configuration-resolution-errors for more details. This behaviour has been deprecated and is scheduled to be removed in Gradle 7.0.

**Comment by mnlipp** [comment]
I'll have a closer look at this next week.

**Comment by mnlipp** [comment]
From my understanding, this is not a problem of mdoclet, but a problem of gradle's javadoc task. There won't be any problem once https://github.com/gradle/gradle/issues/6742 has been fixed. Therefore, I think it makes more sense to wait for this instead of introducing kotlin here only for a workaround.

In case nobody fixes https://github.com/gradle/gradle/issues/6742 in Gradle 7.0, creating a jar with dependencies could also be done using groovy.

**Comment by vlsi** [comment]
Even though the issue is at Gradle side, there's a trivial workaround that can be done at mdoclet side. Why don't you implement a workaround until Gradle implements the full solution?

**Comment by mnlipp** [comment]
Because I don't like your workaround. One point is the usage of kotlin. Now, I don't mind using kotlin in general, but introducing it in a project to implement a workaround makes the project unnecessarily complicated.

Your workaround could also be implemented in Groovy. Which brings me to the main reason: I don't like this artificially created temporary file. Such things are among the dirtiest form of workaround and they tend to cause trouble sooner or later (believe me, I have a long history using "make"). Besides, the workaround is not "trivial", it requires a lot of understanding about the underlying mechanisms.

So, of course, if Gradle 7.0 comes out without a fix, I'll use even the dirtiest workaround to keep things running. But until this happens, I think that the warning should keep "the finger pointed" at the Gradle guys.

**Comment by vlsi** [comment]
> if Gradle 7.0 comes out without a fix

The issue with the current code is it produces a build-type warning, and I prefer to avoid warnings because it is not clear if the warning is produced by the build script itself or by one of the plugins.

**Comment by mnlipp** [comment]
Well, if you want to get rid of the warning, I suggest to submit a PR to Gradle...

**Comment by mnlipp** [comment]
Just tested mdoclet 1.0.5 with gradle 5.6.4 and 6.1.1. There are no more warnings, seems to have been fixed.

**Comment by nedtwigg** [comment]
FWIW, I'm still seeing the warning with gradle 6.2.1.  I made a few small tweaks to vlsi's workaround code here: https://github.com/diffplug/blowdryer-diffplug/blob/5375b4a29b891ca0333142f31b1334a7fe2ad89c/src/main/resources/helper/javadoc-markdown.gradle

**Comment by vlsi** [comment]
>  I made a few small tweaks

It looks like you gave up with configuration avoidance.
Just in case: my WA enables me to skip downloading mlocket jar file in case javadoc is not executed.

**Comment by mnlipp** [comment]
@nedtwigg: Sorry, but I just tested one of my projects with gradle 6.2.1. No warnings for javadoc with mdoclet. (https://github.com/mnlipp/jgrapes/blob/c65ec55aa6da447c33af9a1d8e91c37625b7e21a/org.jgrapes.core/build.gradle#L20)

**Comment by nedtwigg** [comment]
> It looks like you gave up with configuration avoidance.

not intentionally, and I think this approach also allows to skip downloading mdoclet jar, but I make mistakes around that all the time :).  Only reason I made any changes was to resolve mdoclet from the buildscript repos, rather than the project repos.

>  I just tested one of my projects with gradle 6.2.1. No warnings for javadoc with mdoclet.

Hmm, I just cloned that project, which indeed uses 6.2.1, and I'm seeing warnings.  Perhaps you have disabled the warning display somehow?

```
ntwigg@ntwigg-mac-2018 dev % git clone https://github.com/mnlipp/jgrapes/
Cloning into 'jgrapes'...
remote: Enumerating objects: 2933, done.
remote: Counting objects: 100% (2933/2933), done.
remote: Compressing objects: 100% (611/611), done.
remote: Total 341773 (delta 2488), reused 2717 (delta 2272), pack-reused 338840
Receiving objects: 100% (341773/341773), 241.61 MiB | 12.48 MiB/s, done.
Resolving deltas:  57% (168872/296027)
Resolving deltas: 100% (296027/296027), done.
ntwigg@ntwigg-mac-2018 dev % 
ntwigg@ntwigg-mac-2018 dev % cd jgrapes 
ntwigg@ntwigg-mac-2018 jgrapes % ./gradlew javadoc --warning-mode all --stacktrace

> Configure project :
The configuration :examples:org.jdrupes.mdoclet was resolved without accessing the project in a safe manner.  This may happen when a configuration is resolved from a different project. This behaviour has been deprecated and is scheduled to be removed in Gradle 7.0. See https://docs.gradle.org/6.2.1/userguide/viewing_debugging_dependencies.html#sub:resolving-unsafe-configuration-resolution-errors for more details.
   ...

ntwigg@ntwigg-mac-2018 jgrapes % ./gradlew --version

------------------------------------------------------------
Gradle 6.2.1
------------------------------------------------------------
```

**Comment by mnlipp** [comment]
> Perhaps you have disabled the warning display somehow?

Any idea about how I might have done that?

```
$ ./gradlew -Dorg.gradle.java.home=/usr/lib/jvm/java-1.8.0-openjdk javadoc --warning-mode all

> Task :examples:javadoc
/home/xxx/devel/JGrapes/examples/src/org/jgrapes/examples/io/tcpecho/EchoServer.java:56: warning - @startuml is an unknown tag.
/home/xxx/devel/JGrapes/examples/src/org/jgrapes/examples/io/tcpecho/EchoServer.java:56: warning - @enduml is an unknown tag.
/home/xxx/devel/JGrapes/examples/src/org/jgrapes/examples/io/tcpecho/EchoServer2.java:56: warning - @startuml is an unknown tag.
/home/xxx/devel/JGrapes/examples/src/org/jgrapes/examples/io/tcpecho/EchoServer2.java:56: warning - @enduml is an unknown tag.
/home/xxx/devel/JGrapes/examples/src/org/jgrapes/examples/io/consoleinput/EchoUntilQuit.java:51: warning - @startuml is an unknown tag.
/home/xxx/devel/JGrapes/examples/src/org/jgrapes/examples/io/consoleinput/EchoUntilQuit.java:51: warning - @enduml is an unknown tag.
6 warnings

BUILD SUCCESSFUL in 23s
17 actionable tasks: 17 executed
```

**Comment by vlsi** [comment]
Ok. The way to trigger the warning is to add `-Dorg.gradle.parallel=true` or add `org.gradle.parallel=true` to `gradle.properties`.

I can replicate the warning as well with `./gradlew javadoc --warning-mode all --stacktrace -Dorg.gradle.parallel=true` with Java 1.8

PS. It is really sad to spend time on something that is documented.

There are two issues here:
1) Unsafe API is used
2) doclet configuration is resolved even in case user doesn't want to build javadocs

Do we really need to spend time and effort on discussing that?
Really?

**Comment by nedtwigg** [comment]
> Do we really need to spend time and effort on discussing that?

FWIW, I learned a lot from this issue, and especially from your workaround :). And I'm surprised that it's only generated in parallel builds, I didn't realize I had that enabled, so thanks for figuring that out too.  I'm sorry you found this sad, but it helped me.  If a discussion bothers you, it's always fine to unsub or mute it, which is a friendlier thing to do than express disappointment at the rest of us who are catching up.

**Comment by mnlipp** [comment]
I think I have just found the real root cause of the warning. I'll do a bit more testing tomorrow, but I think this can be fixed soon.

**Comment by nedtwigg** [comment]
I don't believe that commit fixed it.

**Comment by vlsi** [comment]
I believe that commit fixed unsafe resolution.
Of course, it did not fix lazy resolution, but who matters?

**Comment by nedtwigg** [comment]
Huh.  I don't care about lazy resolution, so sounds good with me if it fixes the bug, and I think you understand the issue better than I do.  I'll just have to wait for a release to test :D

**Comment by mnlipp** [comment]
>  I'll just have to wait for a release to test :D

I have released it. Due to the change of the gradle version and because Sonatype Nexus was extremely slow once again and failed twice, I had to make several attempts, so the version is 1.0.9.

The warning was about a configuration being resolved from another project. I had always wondered about this, because the plugin creates a configuration which is only used by the project itself. I have no idea how I came up initially with this `project.gradle.afterProject`. But obviously (and not very surprisingly) the closure provided is evaluated in the wrong context (look it up, it's a notification for "any" project having been evaluated, so really not appropriate here). 

By moving the action to `project.afterProject`, I make sure that the closure is "called immediately after a project is evaluated" which apparently implies "from the project context". So resolving the configuration in `project.afterProject` is no longer a "cross-project" resolution and gradle has no problem with it, even in a parallel build.
